apicast/src/plugin.lua

```diff
@@ -0,0 +1,57 @@
+-- Apicast provides a plugin system. To add a custom plugin, we only need to
+-- define a module that implements a few methods like rewrite(), access(), etc.
+-- that are called in the appropriate nginx phases.
+-- The goal of this lua module is to implement the methods that are not likely
+-- to be changed from plugin to plugin so we do not have to repeated code
+-- across all of them.
+
+local balancer = require 'balancer'
+
+-- Plugins need access to provider and configuration, so we need to expose them
+local _M = { provider = require 'provider',
```

Comment:
Then post_action, body_filter and header_filter can be set to nil. And access() can be just a bit duplicated. Or we can try how to extract similar parts of access somewhere else like public table methods.